File: libs/server-nest/persistent/src/persistent/filesystem/workspace-logical-relationships.ts
import { randomUUID } from 'node:crypto';
import { mkdir, rm, writeFile } from 'node:fs/promises';
import { join } from 'node:path';
import {
  DomainError,
  LogicalRelationship,
  type LogicalRelationshipDescription,
  Ref,
  type WorkspaceLogicalRelationships,
} from '@evidence/server-nest-domain';
import { stringify } from 'yaml';
import { EntityList } from '../database';
import {
  listYamlFiles,
  optionalString,
  readYamlRecord,
  requiredString,
  type YamlRecord,
} from './model-files';
import { FileWorkspaceLogicalEntities } from './workspace-logical-entities';

interface RelationshipRecord {
  document: YamlRecord;
  path: string;
  relationship: LogicalRelationship;
}

export class FileWorkspaceLogicalRelationships
  extends EntityList<LogicalRelationship>
  implements WorkspaceLogicalRelationships
{
  private readonly associationsDirectory: string;
  private readonly entities: FileWorkspaceLogicalEntities;

  constructor(
    private readonly workspaceId: string,
    evidenceRoot: string,
  ) {
    super();
    this.associationsDirectory = join(evidenceRoot, 'associations');
    this.entities = new FileWorkspaceLogicalEntities(workspaceId, evidenceRoot);
  }

  override async size(): Promise<number> {
    return (await this.load()).length;
  }

  protected override async findEntities(
    from: number,
    to: number,
  ): Promise<LogicalRelationship[]> {
    return (await this.load())
      .slice(from, to)
      .map((record) => record.relationship);
  }

  protected override async findEntity(
    id: string,
  ): Promise<LogicalRelationship | null> {
    return (await this.findRecord(id))?.relationship ?? null;
  }

  async add(
    desc: LogicalRelationshipDescription,
  ): Promise<LogicalRelationship> {
    await this.validateDescription(desc);
    const id = await this.availableId(desc);
    const document = relationshipDocument(id, desc);
    return this.writeRelationship(
      join(this.associationsDirectory, `${id}.yaml`),
      document,
    );
  }

  async update(
    relationshipId: string,
    desc: LogicalRelationshipDescription,
  ): Promise<LogicalRelationship> {
    const record = await this.findRecord(relationshipId);
    if (!record) {
      throw DomainError.notFound(
        `logical relationship ${relationshipId} not found`,
      );
    }
    await this.validateDescription(desc);
    return this.writeRelationship(record.path, {
      ...record.document,
      id: relationshipId,
      name:
        optionalString(record.document.name) ??
        relationshipName(relationshipId),
      ...(desc.label ? { label: desc.label } : { label: undefined }),
      source: desc.source.id(),
      target: desc.target.id(),
    });
  }

  async delete(relationshipId: string): Promise<void> {
    const record = await this.findRecord(relationshipId);
    if (!record) {
      throw DomainError.notFound(
        `logical relationship ${relationshipId} not found`,
      );
    }
    try {
      await rm(record.path);
    } catch (error) {
      throw DomainError.internal(
        `delete logical relationship file ${record.path}: ${errorMessage(error)}`,
      );
    }
  }

  async list(
    page: number,
    pageSize: number,
  ): Promise<[LogicalRelationship[], number]> {
    rejectInvalidPage(page, pageSize);
    const records = await this.load();
    const offset = (page - 1) * pageSize;
    return [
      records
        .slice(offset, offset + pageSize)
        .map((record) => record.relationship),
      records.length,
    ];
  }

  private async load(): Promise<RelationshipRecord[]> {
    const paths = await listYamlFiles(this.associationsDirectory);
    const records = await Promise.all(
      paths.map((path) => this.readRelationship(path)),
    );
    records.sort((left, right) =>
      left.relationship.identity().localeCompare(right.relationship.identity()),
    );
    return records;
  }

  private async findRecord(id: string): Promise<RelationshipRecord | null> {
    return (
      (await this.load()).find(
        (record) => record.relationship.identity() === id,
      ) ?? null
    );
  }

  private async readRelationship(path: string): Promise<RelationshipRecord> {
    const document = await readYamlRecord(path, 'logical relationship');
    const id = requiredString(document, 'id', path, 'logical relationship');
    requiredString(document, 'name', path, 'logical relationship');
    const source = requiredString(
      document,
      'source',
      path,
      'logical relationship',
    );
    const target = requiredString(
      document,
      'target',
      path,
      'logical relationship',
    );

    return {
      document,
      path,
      relationship: new LogicalRelationship(id, {
        workspace: new Ref(this.workspaceId),
        source: new Ref(source),
        target: new Ref(target),
        label: optionalString(document.label),
      }),
    };
  }

  private async writeRelationship(
    path: string,
    document: YamlRecord,
  ): Promise<LogicalRelationship> {
    const compactDocument = Object.fromEntries(
      Object.entries(document).filter(([, value]) => value !== undefined),
    );
    try {
      await mkdir(this.associationsDirectory, { recursive: true });
      await writeFile(
        path,
        stringify(compactDocument, { lineWidth: 0 }),
        'utf8',
      );
    } catch (error) {
      throw DomainError.internal(
        `write logical relationship file ${path}: ${errorMessage(error)}`,
      );
    }
    return (await this.readRelationship(path)).relationship;
  }

  private async validateDescription(
    desc: LogicalRelationshipDescription,
  ): Promise<void> {
    for (const [label, endpointId] of [
      ['source', desc.source.id()],
      ['target', desc.target.id()],
    ] as const) {
      if (!(await this.entities.findByIdentity(endpointId))) {
        throw DomainError.validation(
          `logical relationship ${label} endpoint ${endpointId} not found in workspace ${this.workspaceId}`,
        );
      }
    }
  }

  private async availableId(
    desc: LogicalRelationshipDescription,
  ): Promise<string> {
    const base =
      normalizeIdentifier(
        desc.label ?? `${desc.source.id()}_${desc.target.id()}`,
      ) ?? randomUUID();
    if (!(await this.findRecord(base))) {
      return base;
    }
    for (;;) {
      const candidate = `${base}_${randomUUID().split('-')[0]}`;
      if (!(await this.findRecord(candidate))) {
        return candidate;
      }
    }
  }
}

function relationshipDocument(
  id: string,
  desc: LogicalRelationshipDescription,
): YamlRecord {
  return {
    id,
    kind: 'association',
    name: relationshipName(id),
    ...(desc.label ? { label: desc.label } : {}),
    source: desc.source.id(),
    target: desc.target.id(),
    relationshipType: 'relates_to',
    direction: 'directed',
  };
}

function relationshipName(id: string): string {
  return id
    .split(/[^a-zA-Z0-9]+/)
    .filter(Boolean)
    .map((part) => `${part[0]?.toUpperCase() ?? ''}${part.slice(1)}`)
    .join('');
}

function normalizeIdentifier(value: string): string | null {
  const normalized = value
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '_')
    .replace(/^_+|_+$/g, '');
  return normalized || null;
}

function rejectInvalidPage(page: number, pageSize: number): void {
  if (page <= 0 || pageSize <= 0) {
    throw DomainError.validation('page and pageSize must be greater than 0');
  }
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

File: libs/server-java/persistent/src/main/java/reengineering/ddd/evidence/persistent/filesystem/WorkspaceLogicalRelationships.java
package reengineering.ddd.evidence.persistent.filesystem;

import io.github.jayclock.smartdomain.core.Ref;
import io.github.jayclock.smartdomain.mybatis.AssociationMapping;
import io.github.jayclock.smartdomain.mybatis.database.EntityList;
import jakarta.inject.Inject;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import reengineering.ddd.evidence.domain.DomainException;
import reengineering.ddd.evidence.domain.description.LogicalRelationshipDescription;
import reengineering.ddd.evidence.domain.model.LogicalRelationship;
import reengineering.ddd.evidence.domain.model.Workspace;
import reengineering.ddd.evidence.persistent.mappers.WorkspacesMapper;

@AssociationMapping(
    entity = Workspace.class,
    field = "logicalRelationships",
    parentIdField = "workspaceId")
public final class WorkspaceLogicalRelationships extends EntityList<String, LogicalRelationship>
    implements Workspace.LogicalRelationships {
  private String workspaceId;
  @Inject private WorkspacesMapper workspaces;

  public WorkspaceLogicalRelationships() {}

  @Override
  protected List<LogicalRelationship> findEntities(int from, int to) {
    List<RelationshipRecord> records = load();
    return records.subList(Math.min(from, records.size()), Math.min(to, records.size())).stream()
        .map(RelationshipRecord::relationship)
        .toList();
  }

  @Override
  protected LogicalRelationship findEntity(String id) {
    RelationshipRecord record = findRecord(id);
    return record == null ? null : record.relationship();
  }

  @Override
  public int size() {
    return load().size();
  }

  @Override
  public LogicalRelationship add(LogicalRelationshipDescription description) {
    validateDescription(description);
    String id = availableId(description);
    return write(
        associationsDirectory().resolve(id + ".yaml"), relationshipDocument(id, description), true);
  }

  @Override
  public LogicalRelationship update(
      String relationshipId, LogicalRelationshipDescription description) {
    RelationshipRecord current = findRecord(relationshipId);
    if (current == null) {
      throw DomainException.notFound("logical relationship " + relationshipId + " not found");
    }
    validateDescription(description);
    Map<String, Object> document = new LinkedHashMap<>(current.document());
    document.put("id", relationshipId);
    document.putIfAbsent("name", relationshipName(relationshipId));
    if (description.label() == null) {
      document.remove("label");
    } else {
      document.put("label", description.label());
    }
    document.put("source", description.source().id());
    document.put("target", description.target().id());
    return write(current.path(), document, false);
  }

  @Override
  public void remove(String relationshipId) {
    RelationshipRecord current = findRecord(relationshipId);
    if (current == null) {
      throw DomainException.notFound("logical relationship " + relationshipId + " not found");
    }
    ModelFiles.delete(current.path(), "logical relationship");
  }

  @Override
  public Workspace.Page<LogicalRelationship> list(int page, int pageSize) {
    validatePage(page, pageSize);
    List<RelationshipRecord> records = load();
    int from = Math.min((page - 1) * pageSize, records.size());
    int to = Math.min(from + pageSize, records.size());
    return new Workspace.Page<>(
        records.subList(from, to).stream().map(RelationshipRecord::relationship).toList(),
        records.size());
  }

  private List<RelationshipRecord> load() {
    return ModelFiles.listYamlFiles(associationsDirectory()).stream()
        .map(this::read)
        .sorted(java.util.Comparator.comparing(record -> record.relationship().getIdentity()))
        .toList();
  }

  private RelationshipRecord findRecord(String id) {
    return load().stream()
        .filter(record -> record.relationship().getIdentity().equals(id))
        .findFirst()
        .orElse(null);
  }

  private RelationshipRecord read(Path path) {
    Map<String, Object> document = ModelFiles.readYaml(path, "logical relationship");
    String id = ModelFiles.requiredString(document, "id", path, "logical relationship");
    ModelFiles.requiredString(document, "name", path, "logical relationship");
    String source = ModelFiles.requiredString(document, "source", path, "logical relationship");
    String target = ModelFiles.requiredString(document, "target", path, "logical relationship");
    return new RelationshipRecord(
        document,
        path,
        new LogicalRelationship(
            id,
            new LogicalRelationshipDescription(
                new Ref<>(workspaceId),
                new Ref<>(source),
                new Ref<>(target),
                ModelFiles.optionalString(document.get("label")))));
  }

  private LogicalRelationship write(Path path, Map<String, Object> document, boolean create) {
    if (create) {
      ModelFiles.writeNewYaml(path, document, "logical relationship");
    } else {
      ModelFiles.replaceYaml(path, document, "logical relationship");
    }
    return read(path).relationship();
  }

  private void validateDescription(LogicalRelationshipDescription description) {
    if (!workspaceId.equals(description.workspace().id())) {
      throw DomainException.validation(
          "logical relationship workspace "
              + description.workspace().id()
              + " does not match scoped workspace "
              + workspaceId);
    }
    WorkspaceLogicalEntities entities = new WorkspaceLogicalEntities(workspaceId, workspaces);
    validateEndpoint("source", description.source().id(), entities);
    validateEndpoint("target", description.target().id(), entities);
  }

  private void validateEndpoint(
      String label, String endpointId, WorkspaceLogicalEntities entities) {
    if (entities.findByIdentity(endpointId).isEmpty()) {
      throw DomainException.validation(
          "logical relationship "
              + label
              + " endpoint "
              + endpointId
              + " not found in workspace "
              + workspaceId);
    }
  }

  private String availableId(LogicalRelationshipDescription description) {
    String requested =
        description.label() == null
            ? description.source().id() + "_" + description.target().id()
            : description.label();
    String normalized = normalizeIdentifier(requested);
    String base = normalized == null ? UUID.randomUUID().toString() : normalized;
    String candidate = base;
    while (findRecord(candidate) != null
        || Files.exists(associationsDirectory().resolve(candidate + ".yaml"))) {
      candidate = base + "_" + UUID.randomUUID().toString().substring(0, 8);
    }
    return candidate;
  }

  private Map<String, Object> relationshipDocument(
      String id, LogicalRelationshipDescription description) {
    Map<String, Object> document = new LinkedHashMap<>();
    document.put("id", id);
    document.put("kind", "association");
    document.put("name", relationshipName(id));
    if (description.label() != null) document.put("label", description.label());
    document.put("source", description.source().id());
    document.put("target", description.target().id());
    document.put("relationshipType", "relates_to");
    document.put("direction", "directed");
    return document;
  }

  private Path associationsDirectory() {
    String root = workspaces.findModelRoot(workspaceId);
    if (root == null) {
      throw DomainException.notFound("workspace " + workspaceId + " not found");
    }
    return Path.of(root).toAbsolutePath().normalize().resolve("associations");
  }

  private static String relationshipName(String id) {
    StringBuilder name = new StringBuilder();
    for (String part : id.split("[^a-zA-Z0-9]+")) {
      if (part.isEmpty()) continue;
      name.append(Character.toUpperCase(part.charAt(0))).append(part.substring(1));
    }
    return name.toString();
  }

  private static String normalizeIdentifier(String value) {
    String normalized =
        value
            .toLowerCase(java.util.Locale.ROOT)
            .replaceAll("[^a-z0-9]+", "_")
            .replaceAll("^_+|_+$", "");
    return normalized.isEmpty() ? null : normalized;
  }

  private static void validatePage(int page, int pageSize) {
    if (page <= 0 || pageSize <= 0) {
      throw DomainException.validation("page and pageSize must be greater than 0");
    }
  }

  private record RelationshipRecord(
      Map<String, Object> document, Path path, LogicalRelationship relationship) {}
}
